allennlp/data/tokenizers/huggingface_transformers_tokenizer.py

```diff
@@ -0,0 +1,430 @@
+from typing import Dict, List, Any
+from overrides import overrides
+import logging
+
+import sys
+import unicodedata
+import re
+import html
+
+from allennlp.data.tokenizers import Token, Tokenizer
+
+from transformers.tokenization_auto import AutoTokenizer
+
+logger = logging.getLogger(__name__)
+
+CONTROL_CHARS = u"".join(
+    chr(c) for c in range(sys.maxunicode + 1) if unicodedata.category(chr(c))[0] == "C"
+)
+WHITESPACES = re.findall(r"\s", u"".join(chr(c) for c in range(sys.maxunicode + 1)), re.UNICODE)
+SPIECE_UNDERLINE = u"▁"
+
+
+@Tokenizer.register("huggingface_transformers")
+class HuggingfaceTransformersTokenizer(Tokenizer):
+    """
+    A ``HuggingfaceTransformersTokenizer`` uses a model from HuggingFace's
+    ``transformers`` library to tokenize some input text.  This often means wordpieces
+    (where ``'AllenNLP is awesome'`` might get split into ``['Allen', '##NL', '##P', 'is',
+    'awesome']``), but it could also use byte-pair encoding, or some other tokenization, depending
+    on the pretrained model that you're using.
+
+    We take a model name as an input parameter, which we will pass to
+    ``AutoTokenizer.from_pretrained``.
+
+    A suggestion for a usage *without* TokenIndexer, in order to fully utilize the Hugging Face's API:
+    1. Tokenize with ``tokenize`` or ``tokenize_with_offsets``
+    2. Encode the tokens with ``encode_plus``, in order to get the appropriate form the pretrained model expects
+    3. Create a LabelsField field with the encoded tokens
+    4. Optioanlly, add fields such as `token_type_ids` and `special_tokens_mask`
+
+    Example:
+        tokens = tokenizer.tokenize(text)
+        encoded_inputs = self._tokenizer.encode_plus([token.text for token in tokens],
+                                    add_special_tokens=True,
+                                    return_token_type_ids=True,
+                                    return_special_tokens_mask=True)
+        fields: Dict[str, Field] = {}
+        fields['tokens'] = LabelsField(encoded_inputs['input_ids'])
```

Comment:
This goes pretty strongly against the way that AllenNLP is designed to work with text inputs.  I think the better solution is to improve our `PretrainedTransformerTokenizer` and related classes with whatever functionality is missing, instead of adding an entirely separate, largely duplicate class that doesn't keep to the design principles of the library.